prow/plugins/jira/jira.go

```diff
@@ -0,0 +1,201 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jira
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/andygrunwald/go-jira"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+	jiraclient "k8s.io/test-infra/prow/jira"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	PluginName = "jira"
+)
+
+var (
+	issueNameRegex = regexp.MustCompile(`\b[a-zA-Z]+-[0-9]+\b`)
+)
+
+func init() {
+	plugins.RegisterGenericCommentHandler(PluginName, handleGenericComment, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, _ []config.OrgRepo) (*pluginhelp.PluginHelp, error) {
+	// The Config field is omitted because this plugin is not configurable.
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The Jira plugin links Pull Requests and Issues to Jira issues",
+	}
+	return pluginHelp, nil
+}
+
+type githubClient interface {
+	EditComment(org, repo string, id int, comment string) error
+	GetIssue(org, repo string, number int) (*github.Issue, error)
+	EditIssue(org, repo string, number int, issue *github.Issue) (*github.Issue, error)
+}
+
+func handleGenericComment(pc plugins.Agent, e github.GenericCommentEvent) error {
+	return handle(pc.JiraClient, pc.GitHubClient, pc.Logger, &e)
+}
+
+func handle(jc jiraclient.Client, ghc githubClient, log *logrus.Entry, e *github.GenericCommentEvent) error {
+	// Nothing to do on deletion
+	if e.Action == github.GenericCommentActionDeleted {
+		return nil
+	}
+
+	issueCandidateNames := issueNameRegex.FindAllString(e.Body, -1)
+	issueCandidateNames = append(issueCandidateNames, issueNameRegex.FindAllString(e.IssueTitle, -1)...)
+	if len(issueCandidateNames) == 0 {
+		return nil
+	}
+
+	referencedIssues := sets.String{}
+	for _, match := range issueCandidateNames {
+		if referencedIssues.Has(match) {
+			continue
+		}
+		_, err := jc.GetIssue(match)
+		if err != nil {
+			if !jiraclient.IsNotFound(err) {
+				log.WithError(err).WithField("Issue", match).Error("Failed to get issue")
+			}
+			continue
+		}
+		referencedIssues.Insert(match)
+	}
+
+	wg := &sync.WaitGroup{}
+	for _, issue := range referencedIssues.List() {
+		wg.Add(1)
+		go func(issue string) {
+			defer wg.Done()
+			if err := upsertGithubLinkToIssue(log, issue, jc, e); err != nil {
+				log.WithField("Issue", issue).WithError(err).Error("Failed to ensure Github link on Jira issue")
+			}
+		}(issue)
+	}
+
+	if err := updateComment(e, referencedIssues.UnsortedList(), jc.JiraURL(), ghc); err != nil {
```

Comment:
Isn't this needlessly called on a title change?

Reply:
I don't think it will even get title changes as it only listens to genericCommentEvents